bittide-rust-ffi/src/lib.rs

```diff
@@ -247,6 +216,86 @@ const _: () = assert!(
     }
 );
 
+fn speed_change_from_ffi(val: &SpeedChange) -> callisto::SpeedChange {
+    match val {
+        SpeedChange::SpeedUp => callisto::SpeedChange::SpeedUp,
+        SpeedChange::SlowDown => callisto::SpeedChange::SlowDown,
+        SpeedChange::NoChange => callisto::SpeedChange::NoChange,
+    }
+}
+
+fn speed_change_to_ffi(val: &callisto::SpeedChange) -> SpeedChange {
+    match val {
+        callisto::SpeedChange::SpeedUp => SpeedChange::SpeedUp,
+        callisto::SpeedChange::SlowDown => SpeedChange::SlowDown,
+        callisto::SpeedChange::NoChange => SpeedChange::NoChange,
+    }
+}
+
+fn reframing_state_from_ffi(val: &ReframingState) -> callisto::ReframingState {
+    match val {
+        ReframingState::Detect => callisto::ReframingState::Detect,
+        ReframingState::Done => callisto::ReframingState::Done,
+        ReframingState::Wait {
+            target_correction,
+            cur_wait_time,
+        } => callisto::ReframingState::Wait {
+            target_correction: *target_correction,
+            cur_wait_time: *cur_wait_time,
+        },
+    }
+}
+
+fn reframing_state_to_ffi(val: &callisto::ReframingState) -> ReframingState {
+    match val {
+        callisto::ReframingState::Detect => ReframingState::Detect,
+        callisto::ReframingState::Done => ReframingState::Done,
+        callisto::ReframingState::Wait {
+            target_correction,
+            cur_wait_time,
+        } => ReframingState::Wait {
+            target_correction: *target_correction,
+            cur_wait_time: *cur_wait_time,
+        },
+    }
+}
+
+fn control_state_from_ffi(st: &ControlSt) -> callisto::ControlSt {
+    return callisto::ControlSt {
+        z_k: st.z_k,
+        b_k: speed_change_from_ffi(&st.b_k),
+        steady_state_target: st.steady_state_target,
+        rf_state: reframing_state_from_ffi(&st.rf_state),
+    };
+}
+
+fn control_state_to_ffi(st: &callisto::ControlSt, rs: &mut ControlSt) {
+    rs.z_k = st.z_k;
+    rs.b_k = speed_change_to_ffi(&st.b_k);
+    rs.steady_state_target = st.steady_state_target;
+    rs.rf_state = reframing_state_to_ffi(&st.rf_state);
+}
+
+fn control_config_from_ffi(cfg: &ControlConfig) -> callisto::ControlConfig {
+    return callisto::ControlConfig {
+        reframing_enabled: cfg.reframing_enabled,
+        wait_time: cfg.wait_time,
+        target_count: cfg.target_count,
+    };
+}
+
+unsafe fn vsi_from_ptr<'a>(ptr: *const ()) -> &'a [callisto::StabilityIndication] {
+    let usize_ptr: *const usize = mem::transmute(ptr);
+    let data_ptr: *const callisto::StabilityIndication = mem::transmute(usize_ptr.offset(1));
+    return mem::transmute(std::slice::from_raw_parts(data_ptr, *usize_ptr));
```

Comment:
You could use `ptr.cast()` here, that should already work because the type annotations in the let bindings. And in the return you don't need the `mem::transmute`, right? The slice should already be the correct type?

Reply:
Good catch. That's a leftover of the VSI wrapper.

Isn't `cast` just the same as using `as`?

Reply:
`as` also allows to change mutability, `.cast` will only change the type!